File: SDKs/Pascal/Mac/FPC/GameResources.pas
unit GameResources;

interface
	uses SysUtils, SGSDK_Core, SGSDK_Font, SGSDK_Audio, SGSDK_Graphics, SGSDK_Input, SGSDK_Physics, SGSDK_MappyLoader;

	procedure LoadResources();
	procedure FreeResources();
	function GameFont(font: String): Font;
	function GameImage(image: String): Bitmap;
	function GameSound(sound: String): SoundEffect;
	function GameMusic(music: String): Music;
	function GameMap(mapName: String): Map;
 
implementation
	var
		_Images: Array of Bitmap;
		_Fonts: Array of Font;
		_Sounds: Array of SoundEffect;
		_Music: Array of Music;
		_Maps: Array of Map;

		_ImagesStr: Array of String;
		_FontsStr: Array of String;
		_SoundsStr: Array of String;
		_MusicStr: Array of String;
		_MapsStr: Array of String;

		_Background: Bitmap;
		_Animation: Sprite;
		_LoadingFont: Font;
		_StartSound: SoundEffect;

	procedure PlaySwinGameIntro();
	var
		i : Integer;
	begin
		PlaySoundEffect(_StartSound);
		for i:= 0 to 55 do
		begin
			DrawBitmap(_Background, 0, 0);
			DrawSprite(_Animation);
			UpdateSprite(_Animation);
			RefreshScreen(60);
			ProcessEvents();			
		end;
		Sleep(1500);
	end;

	procedure ShowLoadingScreen();
	begin
		_Background := LoadBitmap(GetPathToResource('SplashBack.png', ImageResource));
		DrawBitmap(_Background, 0, 0);
		RefreshScreen(60);
		ProcessEvents();

		_Animation := CreateSprite(LoadBitmap(GetPathToResource('SwinGameAni.png', ImageResource)), 4, 14, 712, 184);
		_Animation.xPos := 41;
		_Animation.yPos := 242;
		_LoadingFont := LoadFont(GetPathToResource('cour.ttf', FontResource), 18);
		_StartSound := LoadSoundEffect(GetPathToResource('SwinGameStart.ogg', SoundResource));

		PlaySwinGameIntro();
	end;

	procedure ShowMessage(message: String; number: Integer);
	begin
		DrawText(message, ColorRed, _LoadingFont, 240, 20 + (25 * number));
		RefreshScreen(60);
		ProcessEvents();
	end;

	procedure EndLoadingScreen();
	begin
		ClearScreen();
		RefreshScreen(60);
		FreeFont(_LoadingFont);
		FreeBitmap(_Background);
		FreeSprite(_Animation);
		FreeSoundEffect(_StartSound);
	end;
	
	procedure NewMap(mapName: String);
	begin
		SetLength(_Maps, Length(_Maps) + 1);
		SetLength(_MapsStr, Length(_MapsStr) + 1);
		_Maps[High(_Maps)] := LoadMap(mapName);
		_MapsStr[High(_MapsStr)] := mapName;
	end;
	
	procedure NewFont(fontName, fileName: String; size: Integer);
	begin
		SetLength(_Fonts, Length(_Fonts) + 1);
		SetLength(_FontsStr, Length(_FontsStr) + 1);
		_Fonts[High(_Fonts)] := LoadFont(GetPathToResource(fileName, FontResource), size);
		_FontsStr[High(_FontsStr)] := fontName;
	end;
	
	procedure NewImage(imageName, fileName: String);
	begin
		SetLength(_Images, Length(_Images) + 1);
		SetLength(_ImagesStr, Length(_ImagesStr) + 1);
		_Images[High(_Images)] := LoadBitmap(GetPathToResource(fileName, ImageResource));
		_ImagesStr[High(_ImagesStr)] := imageName;
	end;
	
	procedure NewSound(soundName, fileName: String);
	begin
		SetLength(_Sounds, Length(_Sounds) + 1);
		SetLength(_SoundsStr, Length(_SoundsStr) + 1);
		_Sounds[High(_Sounds)] := LoadSoundEffect(GetPathToResource(fileName, SoundResource));
		_SoundsStr[High(_SoundsStr)] := soundName;
	end;
	
	procedure NewMusic(musicName, fileName: String);
	begin
		SetLength(_Music, Length(_Music) + 1);
		SetLength(_MusicStr, Length(_MusicStr) + 1);
		_Music[High(_Music)] := LoadMusic(GetPathToResource(fileName, SoundResource));
		_MusicStr[High(_MusicStr)] := musicName;
	end;
	
	procedure LoadFonts();
	begin
		NewFont('ArialLarge', 'arial.ttf', 80);
		NewFont('Courier', 'cour.ttf', 16);
	end;

	procedure FreeFonts();
	var
		i: Integer;
	begin
		for i := Low(_Fonts) to High(_Fonts) do
		begin
			FreeFont(_Fonts[i]);
		end;
	end;

	procedure LoadImages();
	begin
		//NewImage('NoImages', 'Ufo.png');
	end;

	procedure FreeImages();
	var
		i: Integer;
	begin
		for i := Low(_Images) to High(_Images) do
		begin
			FreeBitmap(_Images[i]);
		end;
	end;

	procedure LoadSounds();
	begin
		//NewSound('NoSound', 'sound.ogg');
	end;
	
	procedure LoadMaps();
	begin
		//NewMap('test');
	end;
	
	procedure FreeSounds();
	var
		i: Integer;
	begin
		for i := Low(_Sounds) to High(_Sounds) do
		begin
			FreeSoundEffect(_Sounds[i]);
		end;
	end;

	procedure LoadMusics();
	begin
		//NewMusic('NoMusic', 'sound.mp3');
	end;

	procedure FreeMusics();
	var
		i: Integer;
	begin
		for i := Low(_Music) to High(_Music) do
		begin
			FreeMusic(_Music[i]);
		end;
	end;
	
	procedure FreeMaps();
	var
		i: Integer;
	begin
		for i := Low(_Maps) to High(_Maps) do
		begin
			FreeMap(_Maps[i]);
		end;
	end;

	procedure LoadResources();
	var
		oldW, oldH: Integer;
	begin
		oldW := ScreenWidth();
		oldH := ScreenHeight();

		ChangeScreenSize(800, 600);

		//Remove sleeps once "real" game resources are being loaded
		ShowLoadingScreen();
		ShowMessage('Loading fonts...', 0); 
		LoadFonts();
		Sleep(50);

		ShowMessage('Loading images...', 1);
		LoadImages();
		Sleep(50);

		ShowMessage('Loading sounds...', 2);
		LoadSounds();
		Sleep(50);

		ShowMessage('Loading music...', 3);
		LoadMusics();
		Sleep(50);
		
		ShowMessage('Loading maps...', 3);
		LoadMaps();
		Sleep(50);

		ShowMessage('Game loaded...', 4);
		Sleep(50);
		EndLoadingScreen();

		ChangeScreenSize(oldW, oldH);
	end;

	procedure FreeResources();
	begin
		FreeFonts();
		FreeImages();
		FreeMusics();
		FreeSounds();
		FreeMaps();
	end;

	function GameFont(font: String): Font; 
	var
		i: Integer;
	begin
		for i := Low(_FontsStr) to High(_FontsStr) do
		begin
			if _FontsStr[i] = font then begin
				result := _Fonts[i];
				exit;
			end;
		end;
		raise exception.create('Font ' + font + ' does not exist...');
	end;

	function GameImage(image: String): Bitmap;
	var
	i: Integer;
	begin
		for i := Low(_ImagesStr) to High(_ImagesStr) do
		begin
			if _ImagesStr[i] = image then begin
				result := _Images[i];
				exit;
			end;
		end;
		raise exception.create('Image ' + image + ' does not exist...');
	end;

	function GameSound(sound: String): SoundEffect; 
	var
		i: Integer;
	begin
		for i := Low(_SoundsStr) to High(_SoundsStr) do
		begin
			if _SoundsStr[i] = sound then begin
				result := _Sounds[i];
				exit;
			end;
		end;
		raise exception.create('Sound ' + sound + ' does not exist...');
	end;
	
	function GameMap(mapName: String): Map;
	var
		i: Integer;
	begin
		for i := Low(_MapsStr) to High(_MapsStr) do
		begin
			if _MapsStr[i] = mapName then begin
				result := _Maps[i];
				exit;
			end;
		end;
		raise exception.create('Map ' + mapName + ' does not exist...');
	end;
	
	function GameMusic(music: String): Music;
	var
		i: Integer;
	begin
		for i := Low(_MusicStr) to High(_MusicStr) do
		begin
			if _MusicStr[i] = music then begin
				result := _Music[i];
				exit;
			end;
		end;
		raise exception.create('Music ' + music + ' does not exist...');
	end;
end.

File: Tests/Full/VB6/GameResources.cs
using System;
using System.Collections.Generic;
using System.Drawing;

using SwinGameVB;
using Graphics = SwinGameVB.Graphics;
using Bitmap = SwinGameVB.Bitmap;
using Font = SwinGameVB.Fonts;
using FontStyle = SwinGameVB.FontStyle;


namespace Tests
{
    public static class GameResources
    {
		  private static Dictionary<string, Bitmap> _Images = new Dictionary<string, Bitmap>();
		  private static Dictionary<string, Font> _Fonts = new Dictionary<string, Font>();
		  private static Dictionary<string, SoundEffect> _Sounds = new Dictionary<string, SoundEffect>();
		  private static Dictionary<string, Music> _Music = new Dictionary<string, Music>();
		  private static Dictionary<string, Map> _Maps = new Dictionary<string, Map>();

        private static Bitmap _Background;
        private static Bitmap _Animation;
        private static Font _LoadingFont;
        private static SoundEffect _StartSound;


        private static void LoadFonts()
        {
            NewFont("ArialLarge", "arial.ttf", 80);
            NewFont("Courier", "cour.ttf", 15);
            NewFont("CourierLarge", "cour.ttf", 28);
        }

        private static void LoadImages()
        {
            NewImage("BallImage1", "ball.png");
		    NewImage("BallImage2", "ball2.png");
		    NewImage("SmallBall", "ball_small.png");
		    NewImage("Running", "running.png");
		    NewImage("Explosion", "Explosion.png");
		    NewImage("Ship", "ship.png");
		    NewImage("Sea", "sea.png");
			NewImage("BGA", "BackgroundDrawArea.png");
			NewImage("BG", "BackgroundMain.png");
            NewImage("ExplosionBlue", "explosion_blue.jpg");

            NewImage("Frame1", "F01.png");
		    NewImage("Frame2", "F02.png");
		    NewImage("Frame3", "F03.png");
		    NewImage("Frame4", "F04.png");
		    NewImage("Frame5", "F05.png");
		    NewImage("Frame6", "F06.png");
		    NewImage("enShip", "enShip.png");

		    NewTransparentColourImage("BlueExplosion", "explosion_pro.jpg", Color.Black);
            NewTransparentColourImage("BlueExplosion2", "explosion_blue.jpg", Color.Black);
		    for (int i = 0; i < 40; i++)
		    {
                NewTransparentColourImage("Explode_" + Convert.ToString(i), "explode_" + Convert.ToString(i) + ".jpg", Color.Black);
		    }
        }

        private static void LoadSounds()
        {
            NewSound("Shock", "shock.wav");   
        }

        private static void LoadMusic()
        {
            NewMusic("Fast", "Fast.mp3");   
        }

        private static void LoadMaps()
        {
            NewMap("test");
            NewMap("test3");
        }

        public static void LoadResources()
        {
            int width =Consts.Core.ScreenWidth();
            int height =Consts.Core.ScreenHeight();

           Consts.Core.ChangeScreenSize(800, 600);

            ShowLoadingScreen();

            ShowMessage("Loading fonts...", 0); 
		    LoadFonts();
		   Consts.Core.Sleep(50);
		
		    ShowMessage("Loading images...", 1);
		    LoadImages();
		   Consts.Core.Sleep(50);
		
		    ShowMessage("Loading sounds...", 2);
		    LoadSounds();
		   Consts.Core.Sleep(50);
		
		    ShowMessage("Loading music...", 3);
		    LoadMusic();
		   Consts.Core.Sleep(50);

            ShowMessage("Loading maps...", 4);
		    LoadMaps();
		   Consts.Core.Sleep(50);
		
		    //Add game level loading here...
		
		   Consts.Core.Sleep(50);
		    ShowMessage("Game loaded...", 5);
		   Consts.Core.Sleep(50);
		    EndLoadingScreen(width, height);
        }

        public static void ShowLoadingScreen()
        {
            _Background = Consts.Graphics.LoadBitmap(Consts.Core.GetPathToResource("SplashBack.png", ResourceKind.ImageResource));
            Consts.Graphics.DrawBitmap(_Background, 0, 0);
           Consts.Core.RefreshScreen();
           Consts.Core.ProcessEvents();

           _Animation = Consts.Graphics.LoadBitmap(Consts.Core.GetPathToResource("SwinGameAni.png", ResourceKind.ImageResource));
           _LoadingFont = Consts.Text.LoadFont(Consts.Core.GetPathToResource("cour.ttf", ResourceKind.FontResource), 18);
           _StartSound = Consts.Audio.LoadSoundEffect(Consts.Core.GetPathToResource("SwinGameStart.ogg", ResourceKind.SoundResource));

            PlaySwinGameIntro();
        }

        public static void PlaySwinGameIntro()
        {
           Consts.Core.Sleep(300);

           Consts.Audio.PlaySoundEffect(_StartSound);

            for (int i = 0; i < 14; i++)
            {
                Consts.Graphics.DrawBitmap(_Background, 0, 0);
                Consts.Graphics.DrawBitmapPart(_Animation, (i / 7) * 712, (i % 7) * 184, 712, 184, 41, 242);

               Consts.Core.Sleep(67);

               Consts.Core.RefreshScreen();
               Consts.Core.ProcessEvents();
            }

           Consts.Core.Sleep(1500);
        }

        public static void ShowMessage(String message, int number)
        {   
		   Consts.Text.DrawText(message, Color.Red.ToArgb(), _LoadingFont, 240, 20 + (25 * number));
		   Consts.Core.RefreshScreen();
		   Consts.Core.ProcessEvents();
        }

        public static void EndLoadingScreen(int width, int height)
        {
            Consts.Core.ProcessEvents();
            Consts.Core.Sleep(500);
            Consts.Core.ChangeScreenSize(width, height);
            Consts.Graphics.ClearScreen();
           Consts.Core.RefreshScreen();
           Consts.Core.ProcessEvents();

           Consts.Graphics.FreeBitmap(_Animation);
           Consts.Text.FreeFont(_LoadingFont);
            Consts.Graphics.FreeBitmap(_Background);
           Consts.Audio.FreeSoundEffect(_StartSound);
        }

        private static void NewMap(String mapName)
        {
				_Maps.Add(mapName,Consts.MappyLoader.LoadMap(mapName));
        }

        private static void NewFont(String fontName, String filename, int size)
        {
            _Fonts.Add(fontName, Consts.Text.LoadFont(Consts.Core.GetPathToResource(filename, ResourceKind.FontResource), size));
        }

        private static void NewImage(String imageName, String filename)
        {
            _Images.Add(imageName, Consts.Graphics.LoadBitmap(Consts.Core.GetPathToResource(filename, ResourceKind.ImageResource)));
        }

		  private static void NewTransparentColorImage(String imageName, String fileName, Color transColor)
        {
            _Images.Add(imageName, Consts.Graphics.LoadBitmap_Transparent(Consts.Core.GetPathToResource(fileName, ResourceKind.ImageResource), true, transColor.ToArgb()));
        }

        private static void NewTransparentColourImage(String imageName, String fileName, Color transColor)
        {
            NewTransparentColorImage(imageName, fileName, transColor);
        }

        private static void NewSound(String soundName, String filename)
        {
            _Sounds.Add(soundName, Consts.Audio.LoadSoundEffect(Consts.Core.GetPathToResource(filename, ResourceKind.SoundResource)));
        }

        private static void NewMusic(String musicName, String filename)
        {
            _Music.Add(musicName, Consts.Audio.LoadMusic(Consts.Core.GetPathToResource(filename, ResourceKind.SoundResource)));
        }

        private static void FreeFonts()
        {
            foreach(Font f in _Fonts.Values)
            {
               Consts.Text.FreeFont(f);
            }
				_Fonts.Clear();
        }

        private static void FreeImages()
        {
            foreach(Bitmap b in _Images.Values)
            {
                Consts.Graphics.FreeBitmap(b);
            }
				_Images.Clear();
        }

        private static void FreeSounds()
        {
            foreach(SoundEffect ef in _Sounds.Values)
            {
               Consts.Audio.FreeSoundEffect(ef);
            }
				_Sounds.Clear();
        }

        private static void FreeMusic()
        {
            foreach (Music m in _Music.Values)
            {
               Consts.Audio.FreeMusic(m);
            }
				_Music.Clear();
        }

        private static void FreeMaps()
        {
            foreach (Map m in _Maps.Values)
            {
               Consts.MappyLoader.FreeMap(m);
            }
				_Maps.Clear();
        }

        /// <summary>
        /// Frees All Resources
        /// </summary>
        public static void FreeResources()
        {
            FreeFonts();
            FreeImages();
            FreeMusic();
            FreeSounds();
            FreeMaps();
        }

        public static Font GameFont(String font)
        {
            return _Fonts[font];
        }

        public static Bitmap GameImage(String image)
        {
				return _Images[image];
        }

        public static SoundEffect GameSound(String sound)
        {
            return _Sounds[sound];
        }

        public static Music GameMusic(String music)
        {
            return _Music[music];
        }

        public static Map GameMap(String map)
        {
            return _Maps[map];
        }
    }
}
